### PR TITLE
KEYCLOAK-10033 ServletAuthzCacheLifespanAdapterTest stabilized

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/KEYCLOAK-10033

When analysing the logs, I noticed that the APP Server stopped responding. On one side the Auth Server got the `NoHttpResponseException` when calling the logout url, and on the other hand, the HTMLUnit wasn't able to find the username field in the login page.

My theory is that the logout wasn't successful and therefore, the HTML Unit wasn't able to log in. The fix makes sure it gets some response from the App Server. If the response is empty - it retries.

### CODE REVIEW

**Review by mhajas — COMMENTED** [review]
Hello the PR looks good. I have two questions:

1. Should we have also some tests whether request is retried in case `retry-unfinished-requests: true` and it is not in case `retry-unfinished-requests: false`?
2. Have you tried some loop test on Jenkins machine to prove that your fix actually fixed this unstability? I can help you (it is quite easy just provision a machine and then run test in loop) or do it for you if you need.

**Comment by slaskawi** [comment]
> Should we have also some tests whether request is retried in case retry-unfinished-requests: true and it is not in case retry-unfinished-requests: false?

@mhajas Personally, I don't think so. This is why I haven't documented this option. I believe it should only be used for testing. 

> Have you tried some loop test on Jenkins machine to prove that your fix actually fixed this unstability? I can help you (it is quite easy just provision a machine and then run test in loop) or do it for you if you need.

@mhajas If you could teach me how to do it - that would be awesome! Let me sync and chat about this today :) Thanks a lot!

**Comment by mhajas** [comment]
> > Should we have also some tests whether request is retried in case retry-unfinished-requests: true and it is not in case retry-unfinished-requests: false?
> 
> @mhajas Personally, I don't think so. This is why I haven't documented this option. I believe it should only be used for testing.
> 

I see. Well in that case shouldn't we have it somehow restricted to test only usage? To be honest, I am not sure how to approach this, not sure whether we already have some code for testing usage only as part codebase. @hmlnarik WDYT?  

> > Have you tried some loop test on Jenkins machine to prove that your fix actually fixed this unstability? I can help you (it is quite easy just provision a machine and then run test in loop) or do it for you if you need.
> 
> @mhajas If you could teach me how to do it - that would be awesome! Let me sync and chat about this today :) Thanks a lot!

Sure no problem

**Comment by slaskawi** [comment]
> Have you tried some loop test on Jenkins machine to prove that your fix actually fixed this unstability? I can help you (it is quite easy just provision a machine and then run test in loop) or do it for you if you need.

@mhajas I've been running the test in a loop for an hour now. It looks pretty stable.

**Comment by slaskawi** [comment]
> I see. Well in that case shouldn't we have it somehow restricted to test only usage? To be honest, I am not sure how to approach this, not sure whether we already have some code for testing usage only as part codebase. @hmlnarik WDYT?

Just to add some context - in case of `NoHttpResponseException` exception, with `retry-unfinished-requests` turned on, the Keycloak server will retry a failed request. Please note, that `NoHttpResponseException` may happen while receiving the response (in other words, the request was sent, was consumed by the app server and an error occurred while parsing the response). In such a case, the app server will effectively receive a request *twice*. There's no problem with a logout request - nothing bad may happen. But I have no idea about other corner cases.

So to sum it up - I believe this is a helpful option. Maybe it should even be documented (I'm wondering what do you think about that). But it may potentially be risky in case on non-idempotent requests. It should be definitely set to `false` by default and touched only by advanced users/admins.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you @slaskawi for the PR. This way the PR could resolve the issue, yet the impact is, as you correctly mentioned, a potential to issue the same request twice. I am thus not very keen on including the PR as is.

I wonder whether there is any other way for this being resolved, e.g. by using `HttpClientBuilder.evictIdleConnections`? If not, then the option to retry the request would have to be also guarded per connection, and the caller would have to set it explicitly in `HttpContext` for retry to be available.

**Comment by slaskawi** [comment]
> Thank you @slaskawi for the PR. This way the PR could resolve the issue, yet the impact is, as you correctly mentioned, a potential to issue the same request twice. I am thus not very keen on including the PR as is.

@hmlnarik Let me paste a fragment of my comment from the above here:

> Please note, that NoHttpResponseException may happen while receiving the response (in other words, the request was sent, was consumed by the app server and an error occurred while parsing the response). In such a case, the app server will effectively receive a request twice

The case I'm describing the above is the hardest one since it happens when parsing the response. The App (that Keycloak is sending logout request to) just stops responding (due to a very long GC, a process freeze or something similar). From the Keycloak perspective, we have no way of determining if the request (from Keycloak to an app server) was successful or not.

I think there's a second solution we can try - increase a socket timeout. However, from the Linux networking point of view, all sockets were fine (otherwise we would get an error related to  closed sockets (or broken pipe)). I can also try setting a GC time target (maybe `-XX:MaxGCPauseMillis=200`) but I'm not sure if that was the GC that caused it. I believe it's more related to underlying CI infrastructure.

Let me try increasing timeouts in a separate PR.

**Comment by slaskawi** [comment]
@hmlnarik I guess you was referring to https://stackoverflow.com/questions/10558791/apache-httpclient-interim-error-nohttpresponseexception when putting your previous comment? I'm not sure (it was a while ago) but I think I analysed the App server logs and I noticed it stopped responding. So I'm not sure if evicting stale connections will do the trick. But let's check it also...

**Comment by slaskawi** [comment]
@hmlnarik https://github.com/keycloak/keycloak/pull/6205

**Comment by hmlnarik** [comment]
Replaced by #6205
